Refactor ranking metric `map` to be the same as Spark's

loomlike opened the pull request:
### Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
1. Refactor both python and spark ranking metric `map` to be the same as Spark's.
- Rename `map_at_k` to `map` which basically the same as Spark's `meanAveragePrecision`
- Add MAP@K `map_at_k` to be the same as Spark's `meanAveragePrecisionAt`

2. Refactor evaluation tests to use the same data fixtures between python and spark tests.

Note: With this changes, some notebook tests are failing.
I'll need to update the expected values for `map_at_k` metrics for those notebooks, or simply to use `map` that is our previous `map_at_k`.

### Related Issues
<!--- If it fixes an open issue, please link to the issue here. -->
#2002 
#1990 
#1702 

### References
<!--- References would be helpful to understand the changes. -->
<!--- References can be books, links, etc. -->


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [x] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [x] This PR is being made to `staging branch` and not to `main branch`.

loomlike commented:
btw, why draft PR triggers tests? @miguelgfierro @anargyri Can we disable that?

miguelgfierro commented:
> btw, why draft PR triggers tests? @miguelgfierro @anargyri Can we disable that?

This might be a solution: https://github.com/orgs/community/discussions/25722 we need to try it. Issue: #2003 2005

miguelgfierro commented on fix.txt:
```diff
@@ -0,0 +1,12 @@
+# FAILED unit/examples/test_notebooks_python.py::test_sar_deep_dive_runs - papermill.exceptions.PapermillExecutionError:  
```
Here is a trick I use: 
I usually add the errors in the text of the PR and then I go one by one fixing them. 

This also helps other people in the future to fix similar errors.

miguelgfierro replied:
Another thing, since we joined the LF, there is a new action we need to pass: DCO. You noticed that it failed. The solution here: https://github.com/recommenders-team/recommenders/wiki/How-to-sign-commits

anargyri commented:
@loomlike are the test failures related to the values of the metrics? I vaguely recall that we had found some bugs in the Spark implementation (or they were not matching the definitions), hence our own implementations for some of the metrics.

loomlike commented:
> @loomlike are the test failures related to the values of the metrics? I vaguely recall that we had found some bugs in the Spark implementation (or they were not matching the definitions), hence our own implementations for some of the metrics.

I think the error was coming from test group names (as I changed some of the tests names, e.g. test_python_precision -> test_python_precision_at_k). I changed them in the test group lists. Let's see if that works.

Regarding the MAP, Spark added `meanAveragePrecisionAt` since [v3.0.0](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.mllib.evaluation.RankingMetrics.html#pyspark.mllib.evaluation.RankingMetrics.meanAveragePrecisionAt). So I changed our `map_at_k` to be just `map` and added `map_at_k` which implements "at_k" parts for both python and spark utils.

Regarding the definition of MAP vs. MAP@K, actually what we had before was just [MAP based on Wikipedia definition](https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Mean_average_precision) as well as Spark's implementation.

The [community](https://www.kaggle.com/code/nandeshwar/mean-average-precision-map-k-metric-explained-code) seems to follow the following definition:
> AP@k is typically calculated for one user from all recommendations by averaging P@1, P@2, P@3, ... P@k
which the Spark's `meanAveragePrecisionAt` also follows it (and the codes in this PR also follow that definition).

anargyri commented:
Good, as long as the evaluation notebooks give the same metrics as before we should be fine.

miguelgfierro commented on tests/unit/recommenders/evaluation/test_python_evaluation.py:
```diff
@@ -164,10 +150,10 @@ def test_python_rmse(rating_true, rating_pred):
         == 0
     )
-    assert rmse(rating_true, rating_pred) == pytest.approx(7.254309, TOL)
+    assert rmse(rating_true, rating_pred) == target_metrics["rmse"]
```
In other tests we hard code the results, see https://github.com/recommenders-team/recommenders/tree/main/tests#how-to-create-tests-for-the-recommenders-library with the pattern computation equals value.

I would argue it's more clear and explicit

loomlike replied:
I agree with you when an expected value is clear and easy to understand. E.g.
```
# check if the metric is the best number (rmse == 0) if the prediction is perfect, i.e true == prediction.
assert rmse(rating_true, rating_true) == 0
```

I also agree that it is explicit to put the number directly when we also specify the input in the same way, like:
```
output = subtract_two(2.2345, 1.0)  
assert output == 1.2345  # this is explicit

# versus

output = subtract_two(input_fixture_data_a, input_fixture_data_b)
assert output == 1.2345  # our input is already in a blackbox for reusability
```

So, I still prefer using an expected value fixture, the same reason we use the fixture for the input data:
1. By using a single fixture value, we can reuse them across many tests and easy to maintain the test cases with the expected values.
2. We can assure both python and spark implementation produces the same metrics for the same inputs (in addition to the test we directly check both metrics' equalities, `test_spark_python_match`)

What do you think?

miguelgfierro replied:
> I still prefer using an expected value fixture, the same reason we use the fixture for the input data

There are several things to consider here:
- Fixtures are not designed for the results of computations it is designed for data. Because you want to control how the data is loaded (test level, module, or session). For example, maybe we want to load movielens at a module level and not every test because it adds 1 or 2 seconds for every test. A result of a computation doesn't need that, because it is just one number.
- The pattern should be computation == value. Because a common failure is to do a computation1 = computation2, people don't realize that computation2 can also fail, so then they think they are testing computation1, but the error can come from computation1 or computation2. I think it is important to make clear this to less experienced developers (after all, we are trying to teach good coding practices).
- I agree with you that it is convenient to have common results for python and spark, but then it shouldn´t be a fixture, but constants. We can add them as constants or as a dict, or we can directly add the number to keep the consistency with the rest of the repo. 

Thoughts?

loomlike replied:
@miguelgfierro per your suggestion, I changed the way the tests check values. Please review again and see if this PR is good to merge!

miguelgfierro commented on tests/unit/recommenders/evaluation/test_python_evaluation.py:
```diff
@@ -100,12 +67,31 @@ def rating_pred_binary(rating_pred):
-    # Change data types of true and prediction data, and there should type error produced
-    rating_true[DEFAULT_USER_COL] = rating_true[DEFAULT_USER_COL].astype(str)
-    rating_true[DEFAULT_RATING_COL] = rating_true[DEFAULT_RATING_COL].astype(str)
+    mocked_fn = Mock(return_value=True)
```
A question here, what is the function of the mock here?

loomlike replied:
that is to test if `_check_column_dtypes` returns the same output as the input function when there's no exceptions.
I'll add a comment to the test.

miguelgfierro commented:
@loomlike if you get an error on DCO you can go to details and approve it. Only admins can do this. I've approved this one.

miguelgfierro reviewed (dismissed):
@loomlike hey we got an internal security email and it seems that it is due to an unsecured version of TF. 

That caused a lot of problems, as expected, some of the TF algos with the v1 API break. I tried to figure out a way to fix it, but I couldn't. See https://github.com/recommenders-team/recommenders/pull/2017

Until this is merged to main, we can't start new VMs for the tests because they will have the old TF version and will trigger the security issue. 

So would you mind stop doing any new commit to this PR until the TF situation is fixed?

TF=💩

miguelgfierro commented:
@loomlike the issue is fixed https://github.com/recommenders-team/recommenders/pull/2017

Can you please merge the latest staging into your branch before committing?

miguelgfierro reviewed (approved):
Awesome!
